crates/commitlog/src/stream/writer.rs

```diff
@@ -0,0 +1,428 @@
+                break;
+            };
+
+            let mut current_segment = if buf.starts_with(&segment::MAGIC) {
+                // Ensure the previous segment, if any, is fsync'ed.
+                self.close_current_segment().await?;
+                // Ensure we actually have a valid segment header.
+                let header =
+                    segment::Header::decode(buf).inspect_err(|e| warn!("failed to decode segment header: {e}"))?;
+                trace!(
+                    "create segment at {}",
+                    self.last_written_tx_range
+                        .as_ref()
+                        .map(|range| range.end)
+                        .unwrap_or_default()
+                );
+                let (mut segment, index) = spawn_blocking({
+                    let repo = self.repo.clone();
+                    let last_written_tx_range = self.last_written_tx_range.clone();
+                    let commitlog_options = self.commitlog_options;
+                    move || create_segment(repo, last_written_tx_range, commitlog_options)
+                })
+                .await
+                .unwrap()
+                .map(|(segment, index)| (segment.into_async_writer(), index))?;
+
+                segment.write_all(&buf[..segment::Header::LEN]).await?;
+                stream.consume(segment::Header::LEN as _);
+
+                CurrentSegment {
+                    header,
+                    segment,
+                    offset_index: index,
+                }
+            } else if let Some(current_segment) = self.current_segment.take() {
+                current_segment
+            } else {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "no current segment, expected segment header",
+                ));
+            };
+
+            // What follows is commits to be written to `current_segment`,
+            // until we encounter EOF or a segment marker.
+            let res = self
+                .append_all_inner(&mut stream, &mut current_segment, &mut progress)
+                .await;
+            // Ensure we flush application buffers (BufWriter).
+            // `fsync` can be delayed until we either open a new segment,
+            // or the stream is exhausted and we break the loop.
+            // If there are errors writing to the segment, there's no point in
+            // syncing.
+            current_segment.segment.flush().await?;
+            let maybe_eof = res?;
+            // Put back segment, so it gets closed if we break the loop.
+            self.current_segment = Some(current_segment);
+            match maybe_eof {
+                AppendInnerResult::StreamExhausted => break,
+                AppendInnerResult::SegmentMarker => continue,
+            }
+        }
+
+        if let Some(current_segment) = self.current_segment.as_mut() {
+            current_segment.flush_and_sync().await?;
```

Comment:
I have a ongoing PR where I am not fsyncing after every stream end - https://github.com/clockworklabs/SpacetimeDBPrivate/pull/1452/files#diff-0d8c22747cef364d129ae36bac70d4e6a304385c14d7fffdcfe75a400ca15ef9L727. 

Reason being: most of the stream could be empty or really small for replication, can we make it caller responsibility to fsync and only flush here.

Reply:
The only issue is that we can't do much if the writer is (accidentally) dropped, due to the "async drop" problem. Let me see if I can come up with something.

Reply:
FYI: For replication, I was solving it by explicitly calling fsync on revceiving decide event. https://github.com/clockworklabs/SpacetimeDBPrivate/pull/1452/files#diff-465b61f1322cc15cd6d3bf87a373f0f1d0a20003846b857e842d1b538eb0ec1fR611